src/components/implementation/nicmgr/dpdk/init.c

```diff
@@ -1,20 +1,187 @@
 #include <llprint.h>
 #include <cos_dpdk.h>
 #include <nic.h>
+#include <shm_bm.h>
+#include <netshmem.h>
+#include <sched.h>
+#include <arpa/inet.h>
+
+#include "interface_impl.h"
 
 #define NB_RX_DESC_DEFAULT 1024
 #define NB_TX_DESC_DEFAULT 1024
+
 #define MAX_PKT_BURST 512
 
+#define NIPQUAD(addr) \
+((unsigned char *)&addr)[0], \
+((unsigned char *)&addr)[1], \
+((unsigned char *)&addr)[2], \
+((unsigned char *)&addr)[3]
+#define NIPQUAD_FMT "%u.%u.%u.%u"
+
+struct eth_hdr {
+	u8_t dst_addr[6];
+	u8_t src_addr[6];
+	u16_t ether_type;
+} __attribute__((packed));
+
+struct ether_addr {
+	uint8_t addr_bytes[6]; /**< Addr bytes in tx order */
+} __attribute__((packed));
+
+struct arp_ipv4 {
+	struct ether_addr arp_sha;  /**< sender hardware address */
+	uint32_t          arp_sip;  /**< sender IP address */
+	struct ether_addr arp_tha;  /**< target hardware address */
+	uint32_t          arp_tip;  /**< target IP address */
+} __attribute__((packed));
+
+struct arp_hdr {
+	uint16_t arp_hardware;    /* format of hardware address */
+#define RTE_ARP_HRD_ETHER     1  /* ARP Ethernet address format */
+
+	uint16_t arp_protocol;    /* format of protocol address */
+	uint8_t  arp_hlen;    /* length of hardware address */
+	uint8_t  arp_plen;    /* length of protocol address */
+	uint16_t arp_opcode;     /* ARP opcode (command) */
+#define	RTE_ARP_OP_REQUEST    1 /* request to resolve address */
+#define	RTE_ARP_OP_REPLY      2 /* response to previous request */
+#define	RTE_ARP_OP_REVREQUEST 3 /* request proto addr given hardware */
+#define	RTE_ARP_OP_REVREPLY   4 /* response giving protocol address */
+#define	RTE_ARP_OP_INVREQUEST 8 /* request to identify peer */
+#define	RTE_ARP_OP_INVREPLY   9 /* response identifying peer */
+
+	struct arp_ipv4 arp_data;
+} __attribute__((packed));
+
+struct ip_hdr {
+	u8_t ihl:4;
+	u8_t version:4;
+	u8_t tos;
+	u16_t total_len;
+	u16_t id;
+	u16_t frag_off;
+	u8_t ttl;
+	u8_t proto;
+	u16_t checksum;
+	u32_t src_addr;
+	u32_t dst_addr;
+} __attribute__((packed));
+
+struct tcp_udp_port
+{
+	u16_t src_port;
+	u16_t dst_port;
+};
+
+char* g_rx_mp = NULL;
+char* g_tx_mp = NULL;
+
 static u16_t nic_ports = 0;
 
 static void
-process_packets(cos_portid_t port_id, char** rx_pkts, uint16_t nb_pkts)
+tx_ringbuf_init()
+{
+	vaddr_t buf_addr = NULL;
+
+	buf_addr = malloc(TX_PKT_RING_SZ);
+	assert(buf_addr);
+
+	ck_ring_init(buf_addr, TX_PKT_RBUF_SZ);
+
+	g_tx_ring    = buf_addr;
+	g_tx_ringbuf = buf_addr + sizeof(struct ck_ring);
+}
+
+static struct client_session *
+find_session(uint32_t dst_ip, uint16_t dst_port)
 {
 	int i;
+	for (i = 0; i < NIC_MAX_SESSION; i++) {
```

Comment:
Seems you iterate every session while doing a search. It is good for now but may result in additional overheads. Using hash map may be another solution.